### PR TITLE
[CallPromotionUtil] See through function alias when devirtualizing a virtual call on an alloca.

### PR DESCRIPTION
**Pull request by mingmingl-llvm**
- Extract utility function from [DevirtModule::tryFindVirtualCallTargets](https://github.com/llvm/llvm-project/blob/b99163fe8feeacba7797d5479bbcd5d8f327dd2d/llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp#L1069-L1082), which sees through an alias to a function. Call this utility function in the WPD callsite.
- For type profiling work, this helper function will be used by indirect-call-promotion pass to find the function pointer at a specified vtable offset (an example in [this line](https://github.com/minglotus-6/llvm-project/blob/77a0ef12de82d11f448f7f9de6f2dcf87d9b74af/llvm/lib/Transforms/Instrumentation/IndirectCallPromotion.cpp#L347))

### CODE REVIEW

File: llvm/include/llvm/Analysis/TypeMetadataUtils.h [hunk 1]
```diff
@@ -14,6 +14,7 @@
 #ifndef LLVM_ANALYSIS_TYPEMETADATAUTILS_H
 #define LLVM_ANALYSIS_TYPEMETADATAUTILS_H
 
+#include "llvm/IR/GlobalVariable.h"
```
Comment (teresajohnson) [inline llvm/include/llvm/Analysis/TypeMetadataUtils.h:17]: Can you instead just use a forward declaration for GlobalVariable, like we do for other types below?

File: llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp [hunk 2]
```diff
@@ -1071,12 +1071,10 @@ bool DevirtModule::tryFindVirtualCallTargets(
     if (!Ptr)
```
Comment (teresajohnson) [inline llvm/lib/Transforms/IPO/WholeProgramDevirt.cpp:1071]: Shouldn't this earlier call to getPointerAtOffset be removed now?

File: llvm/include/llvm/Analysis/TypeMetadataUtils.h [hunk 3]
```diff
@@ -77,6 +78,10 @@ void findDevirtualizableCallsForTypeCheckedLoad(
 Constant *getPointerAtOffset(Constant *I, uint64_t Offset, Module &M,
                              Constant *TopLevelGlobal = nullptr);
 
+// Given a vtable, returns the function pointer specified by Offset.
```
Comment (teresajohnson) [inline llvm/include/llvm/Analysis/TypeMetadataUtils.h:81]: Document the pair contents.

**Review by teresajohnson — COMMENTED** [review]
Suggest making the PR title shorter and moving some of the details there currently to the description. Also, I think this can be marked NFC?

**Comment by minglotus-6** [comment]
> Suggest making the PR title shorter and moving some of the details there currently to the description. Also, I think this can be marked NFC?

Makes sense to make PR title shorter and move implementation details. It's not strictly an NFC (not stated in a simple way in prior title), rephrased the title.

**Review by teresajohnson — APPROVED** [review]

**Pull request merged**